File: atb_cpp_qwen3vl_embedding/tests/test_consistency.py
"""
C++ vs Python consistency test.

Loads the C++ engine output from /tmp/cpp_embedding.bin and compares
it against the Python ATB engine output using cosine similarity.

Prerequisites:
    1. Run ./test_consistency (C++ side) first to generate /tmp/cpp_embedding.bin
    2. Model checkpoint at /mnt/workspace/gitCode/models/Qwen3-VL-Embedding-2B/

Usage:
    python tests/test_consistency.py
"""

import struct
import sys
import numpy as np
import torch
import torch.nn.functional as F

# ── Configuration ────────────────────────────────────────────────
CPP_OUTPUT_PATH = "/tmp/cpp_embedding.bin"
MODEL_DIR = "/mnt/workspace/gitCode/models/Qwen3-VL-Embedding-2B"
COSINE_THRESHOLD = 0.99


def load_cpp_embedding(path: str) -> tuple:
    """Load C++ embedding from binary file.

    Returns (dim: int, data: np.ndarray[float32])
    """
    with open(path, "rb") as f:
        dim = struct.unpack("q", f.read(8))[0]
        data = np.frombuffer(f.read(dim * 4), dtype=np.float32)
    return dim, data


def run_python_engine() -> torch.Tensor:
    """Run Python ATB engine with the same input and return embedding."""
    # Set buffer size before importing engine
    sys.path.insert(0, "/mnt/workspace/gitCode/atb_llm")
    from atb_python_qwen3vl_embedding.utils import set_atb_buffer_size
    set_atb_buffer_size(10 * 1024 * 1024 * 1024)

    from atb_python_qwen3vl_embedding.engine import Qwen3VLEngine

    engine = Qwen3VLEngine(MODEL_DIR)

    # Same input_ids as C++ test: [151643, 15339, 1879]
    input_ids = torch.tensor([[151643, 15339, 1879]], dtype=torch.long)

    # encode() does: forward -> pool last token -> L2 normalize
    result = engine.encode(input_ids, normalize=True)
    return result.flatten()


def main():
    print("=" * 60)
    print("C++ vs Python Consistency Test")
    print("=" * 60)

    # ── Load C++ output ──────────────────────────────────────────
    print(f"\n[1] Loading C++ embedding from {CPP_OUTPUT_PATH}")
    try:
        cpp_dim, cpp_data = load_cpp_embedding(CPP_OUTPUT_PATH)
    except FileNotFoundError:
        print(f"ERROR: {CPP_OUTPUT_PATH} not found.")
        print("Run the C++ test first: ./test_consistency")
        sys.exit(1)

    cpp_emb = torch.from_numpy(cpp_data).float()
    print(f"    C++ embedding dim: {cpp_dim}")

    # ── Run Python engine ────────────────────────────────────────
    print(f"\n[2] Running Python ATB engine")
    py_emb = run_python_engine().float()
    print(f"    Python embedding dim: {py_emb.shape[0]}")

    # ── Compare ──────────────────────────────────────────────────
    print(f"\n[3] Comparison:")
    print(f"    dim: C++={cpp_dim}, Python={py_emb.shape[0]}")

    if cpp_dim != py_emb.shape[0]:
        print(f"FAIL: dimension mismatch")
        sys.exit(1)

    cos = F.cosine_similarity(cpp_emb, py_emb, dim=0).item()
    mse = F.mse_loss(cpp_emb, py_emb).item()
    max_diff = (cpp_emb - py_emb).abs().max().item()

    print(f"    cosine similarity: {cos:.6f}")
    print(f"    MSE:               {mse:.2e}")
    print(f"    max abs diff:      {max_diff:.2e}")

    # ── Print first 8 values ─────────────────────────────────────
    print(f"\n[4] First 8 values:")
    cpp_first8 = cpp_data[:8].tolist()
    py_first8 = py_emb[:8].tolist()
    print(f"    C++:    {[f'{v:.6f}' for v in cpp_first8]}")
    print(f"    Python: {[f'{v:.6f}' for v in py_first8]}")

    # ── Verdict ──────────────────────────────────────────────────
    print(f"\n{'=' * 60}")
    if cos > COSINE_THRESHOLD:
        print(f"PASS: cosine similarity {cos:.6f} > {COSINE_THRESHOLD}")
        sys.exit(0)
    else:
        print(f"FAIL: cosine similarity {cos:.6f} < {COSINE_THRESHOLD}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: atb_cpp_qwen3vl_embedding/tests/test_consistency.cpp
/**
 * C++ vs Python consistency test.
 *
 * Runs text-only inference with fixed input_ids and saves the output
 * embedding as a binary file for comparison with the Python engine.
 *
 * Binary format:
 *   [int64_t] hidden_dim
 *   [float32 * hidden_dim] embedding data (converted from fp16)
 *
 * Run: ./test_consistency
 * Then: python tests/test_consistency.py
 */

#include "atb_llm/types.h"
#include "atb_llm/engine.h"
#include "utils/float_utils.h"
#include "log/logger.h"

#include <cstdio>
#include <cstring>
#include <vector>
#include <cmath>
#include <cstdint>

#define IS_OK(s) ((s) == atb_llm::STATUS_OK)

static const char* MODEL_DIR = "/mnt/workspace/gitCode/models/Qwen3-VL-Embedding-2B";
static const char* OUTPUT_PATH = "/tmp/cpp_embedding.bin";

int main(int argc, char** argv) {
    LOG_INFO("=== C++ vs Python Consistency Test ===");

    // ── 1. Create engine ─────────────────────────────────────
    atb_llm::EngineConfig config;
    config.model_dir = MODEL_DIR;
    config.buffer_size = 10LL * 1024 * 1024 * 1024;  // 10GB
    config.device_id = 0;
    config.normalize = true;  // L2-normalize output

    std::unique_ptr<atb_llm::LLMEngine> engine;
    atb_llm::Status s = atb_llm::LLMEngine::Create(config, engine);
    if (!IS_OK(s) || !engine) {
        LOG_ERROR("Failed to create engine: %d", static_cast<int>(s));
        return 1;
    }
    LOG_INFO("Engine created successfully");

    // ── 2. Prepare input ─────────────────────────────────────
    // Fixed input: token IDs [151643, 15339, 1879] (seq_len=3)
    int64_t input_ids[] = {151643, 15339, 1879};
    int64_t seq_len = 3;

    atb_llm::InferRequest request;
    request.mode = atb_llm::InputMode::TEXT_ONLY;
    request.text.input_ids = input_ids;
    request.text.batch_size = 1;
    request.text.seq_length = seq_len;

    // ── 3. Run inference ─────────────────────────────────────
    atb_llm::InferResult result;
    s = engine->Encode(request, result);
    if (!IS_OK(s)) {
        LOG_ERROR("Encode failed: %d", static_cast<int>(s));
        return 1;
    }

    LOG_INFO("Inference succeeded: shape=(%ld), dtype=%d",
             static_cast<long>(result.shape[0]), static_cast<int>(result.dtype));

    // ── 4. Convert to float32 ────────────────────────────────
    int64_t dim = result.shape[0];
    std::vector<float> fp32_data(dim);

    if (result.dtype == ACL_FLOAT16) {
        const uint16_t* fp16_data = result.As<uint16_t>();
        for (int64_t i = 0; i < dim; i++) {
            fp32_data[i] = atb_llm::Fp16ToF32(fp16_data[i]);
        }
        LOG_INFO("Converted fp16 -> fp32 (%ld elements)", static_cast<long>(dim));
    } else if (result.dtype == ACL_FLOAT) {
        const float* f32_data = result.As<float>();
        std::memcpy(fp32_data.data(), f32_data, dim * sizeof(float));
        LOG_INFO("Output already fp32 (%ld elements)", static_cast<long>(dim));
    } else {
        LOG_ERROR("Unexpected dtype: %d", static_cast<int>(result.dtype));
        return 1;
    }

    // ── 5. Print first 8 values ──────────────────────────────
    LOG_INFO("First 8 embedding values:");
    for (int i = 0; i < 8 && i < static_cast<int>(dim); i++) {
        fprintf(stderr, "  [%d] = %.6f\n", i, fp32_data[i]);
    }

    // ── 6. Save to binary file ───────────────────────────────
    FILE* f = fopen(OUTPUT_PATH, "wb");
    if (!f) {
        LOG_ERROR("Failed to open output file: %s", OUTPUT_PATH);
        return 1;
    }

    // Write header: int64_t dimension
    fwrite(&dim, sizeof(int64_t), 1, f);
    // Write data: float32 embedding
    fwrite(fp32_data.data(), sizeof(float), dim, f);
    fclose(f);

    LOG_INFO("Saved embedding to %s (dim=%ld, %ld bytes)",
             OUTPUT_PATH, static_cast<long>(dim),
             static_cast<long>(sizeof(int64_t) + dim * sizeof(float)));

    LOG_INFO("=== Consistency test C++ side complete ===");
    LOG_INFO("Run: python tests/test_consistency.py");

    return 0;
}
